synapse/handlers/event_auth.py

```diff
@@ -160,6 +160,27 @@ async def check_host_in_room(self, room_id: str, host: str) -> bool:
         with Measure(self._clock, "check_host_in_room"):
             return await self._store.is_host_joined(room_id, host)
 
+    async def assert_host_in_room(
+        self, room_id: str, host: str, allow_partial_state_rooms: bool = False
+    ) -> None:
+        """
+        Asserts that the host is in the room, or raises an AuthError.
+
+        If the room is partial-stated, we raise an AuthError with the
+        UNABLE_DUE_TO_PARTIAL_STATE flag, unless `allow_partial_state_rooms` is true.
+        """
+        if not allow_partial_state_rooms and await self._store.is_partial_state_room(
+            room_id
+        ):
+            raise AuthError(
+                403,
+                "Unable to authorise you right now; room is partial-stated here.",
+                errcode=Codes.UNABLE_DUE_TO_PARTIAL_STATE,
+            )
```

Comment:
unsure if 403 is right here, maybe 503?

(If "authorise" is the right word here, I think 401 might make more sense than 403?)

Reply:
Hmm, I used 403 because that's what we use for 'host not in room' and the semantics are quite similar.

503 would be 'service unavailable' and IIRC it's one of the codes that lands us in hot water with CloudFlare so we shouldn't raise that.

Reply:
My reading of https://support.cloudflare.com/hc/en-us/articles/115003011431-Troubleshooting-Cloudflare-5XX-errors#502504error was this the codes are 502 and 504, but I can easily believe 503 also triggers some unwanted behaviour on our end.

In any case, you've thought about it and your reasoning seems sound to me. Thanks for indulging me.